database/json_file_storage.go

```diff
@@ -0,0 +1,175 @@
+	logger  logging.Logger
+}
+
+// NewJSONFileStorage creates a new JSONFileStorage instance
+func NewJSONFileStorage(logger logging.Logger, dir string, networks []ids.ID) (*JSONFileStorage, error) {
+	storage := &JSONFileStorage{
+		dir:     filepath.Clean(dir),
+		mutexes: make(map[ids.ID]*sync.RWMutex),
+		logger:  logger,
+	}
+
+	for _, network := range networks {
+		storage.mutexes[network] = &sync.RWMutex{}
+	}
+
+	_, err := os.Stat(dir)
+	if err == nil {
+		// dir already exist
+		// return the existing storage
+		return storage, nil
+	}
+
+	// 0755: The owner can read, write, execute.
+	// Everyone else can read and execute but not modify the file.
+	err = os.MkdirAll(dir, 0755)
+	if err != nil {
+		storage.logger.Error("failed to create directory",
+			zap.String("dir", dir),
+			zap.Error(err))
+		return nil, err
+	}
+
+	return storage, nil
+}
+
+// Get the latest chain state from the json database, and retrieve the value from the key
+func (s *JSONFileStorage) Get(chainID ids.ID, key []byte) ([]byte, error) {
+	currentState := make(chainState)
+	fileExists, err := s.read(chainID, &currentState)
+	if err != nil {
+		s.logger.Error(
+			"failed to read file",
+			zap.String("chainID", chainID.String()),
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	if !fileExists {
+		return nil, errors.Errorf("file does not exist. chainID: %s", chainID)
+	}
+
+	if _, ok := currentState[string(key)]; !ok {
+		return nil, errors.Errorf("key does not exist. chainID: %s key: %s", chainID, key)
+	}
+
+	return []byte(currentState[string(key)]), nil
+}
+
+// Put the value into the json database. Read the current chain state and overwrite the key, if it exists
+func (s *JSONFileStorage) Put(chainID ids.ID, key []byte, value []byte) error {
+	currentState := make(chainState)
+	_, err := s.read(chainID, &currentState)
+	if err != nil {
+		s.logger.Error(
+			"failed to read file",
+			zap.String("chainID", chainID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	currentState[string(key)] = string(value)
+
+	return s.write(chainID, currentState)
+}
+
+// write value into the file
+func (s *JSONFileStorage) write(network ids.ID, v interface{}) error {
+	mutex, ok := s.mutexes[network]
+	if !ok {
+		return errors.Errorf("network does not exist. chainID: %s", network.String())
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	fnlPath := filepath.Join(s.dir, network.String()+".json")
+	tmpPath := fnlPath + ".tmp"
+
+	b, err := json.MarshalIndent(v, "", "\t")
+	if err != nil {
+		return err
+	}
+
+	// write marshaled data to the temp file
+	// if write failed, the original file is not affected
+	// 0644 Only the owner can read and write.
+	// Everyone else can only read. No one can execute the file.
+	if err := os.WriteFile(tmpPath, b, 0644); err != nil {
+		return errors.Wrap(err, "failed to write file")
+	}
+
+	// move final file into place
+	if err := os.Rename(tmpPath, fnlPath); err != nil {
+		return errors.Wrap(err, "failed to rename file")
+	}
+
```

Comment:
nit: We may want to consider to performing a sync on the file if we want to be sure it is on disk (and not just in the writeback buffer of the file cache). Not sure if it is necessary in our use case.

Reply:
I wasn't aware of this until you pointed it out, but Go `File`s do not perform buffered writes, so as soon as `File.Write` is called (as is the case internally in all of the `os` function calls here), the changes are written to disk. See here for a discussion on this: https://stackoverflow.com/a/10890222

Reply:
I suspect golang's File object is a light wrapper around a file descriptor (created using an _open_ syscall). Writes to it are not buffered at the application level (unlike files opened with _fopen_) so they do not need to be flushed, but they are still written to the file system's buffercache first and later written back to disk.

An explicit sync is not necessary if the system does not abruptly power down. However, if there is a power outage, data in the buffercache that hasn't been written to disk can be lost. Most applications don't bother with calling sync as it hurts performance. However, databases often do, and we are building a type of database here.

With that said, a sync here might not be necessary if we assume that the filesystem will writeback the inode update (for the rename) after the writes to the file. This is probably a good assumption for most filesystems. We may also want to consider adding a checksum to the file to make sure we never use corrupted data.